### PR TITLE
Improve CI speed

### PR DESCRIPTION
**Pull request by akrantz01**
### Summary
Speeds up the total runtime of the CI by 200% at the minimum through parallelization and caching.
| Job                               | Old runtime | New runtime | Percent change |
|-----------------------------------|-------------|-------------|----------------|
| Check Formatting                  | 1m 30s      | 20s         | -450%          |
| Check STM32CubeMX Code Generation | 3m 25s      | 50s         | -410%          |
| Build and Run x86 Tests           | 7m 30s      | 1m 50s      | -409%          |
| Build ARM Executables             | 7m 20s      | 2m 40s      | -275%          |

This is a long pull request; however, a good portion of the additions are from the `Pipfile.lock`, which is a generated file, and a good part of the removals are from unused Travis CI scripts, which wouldn't have many benefits being in a separate PR. Bumping the copyright dates on the STM32CubeMX generated files could probably be done in a separate PR, but I just needed to verify that everything worked.

### Changelist 
- splits format checking, codegen checking, ARM builds, and x86 tests into separate jobs
- adds caching for compilation with [ccache](https://ccache.dev/)
  - statistics on cache use are reported at the end of each run
- adds caching for Python dependencies
- parallelizes running tests in debug and release mode
- parallelizes code generation from the STM32CubeMX files for each board
- parallelizes building ARM executables for each board
- saves the generated ARM executables as artifacts available for download

### Testing Done
The new CI should be functionally equivalent to the current one.

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer than **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -54,7 +54,6 @@
 *.ini
 
 # Python
-Pipfile.lock
```
Comment (pasitlaothamatas) [inline .gitignore:57]: @akrantz01 Why isn't Pipfile.lock no longer in the .gitignore?

Comment (pasitlaothamatas) [reply .gitignore:57]: oh i see this has something to do with cachecing?

**Comment by pasitlaothamatas** [comment]
is there a way to make the jobs appear like this? i think its a little bit neater
image

File: .github/workflows/githubactions.yml [hunk 2]
```diff
@@ -1,87 +1,258 @@
...
+        run: git --no-pager diff --color --exit-code
+
+  tests:
+    name: Build and Run x86 Tests
+    runs-on: ubuntu-18.04
+    strategy:
+      fail-fast: false
+      matrix:
+        mode:
+          - debug
```
Comment (pasitlaothamatas) [inline .github/workflows/githubactions.yml:99]: as discussed on slack, no need to split into debug and release :)

**Review by timbitss — APPROVED** [review]
lgtm

File: .github/workflows/githubactions.yml [hunk 3]
```diff
@@ -1,87 +1,258 @@
...
+      - name: Setup CCache
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.CCACHE_DIR }}
+          key: ccache-${{ matrix.board }}-build
+
+      - name: Install ARM GCC
+        env:
+          # Unfortunately, the download link is a literal URL which means it does not
+          # pull the latest version automatically.
```
Comment (pasitlaothamatas) [inline .github/workflows/githubactions.yml:210]: probably best that the version (arm-none-eabi-9-2019-q4-major) of this is the same as our embedded builds? doesnt need to be the latest version

File: boards/FSM/Inc/stm32f3xx_hal_conf.h [hunk 4]
```diff
@@ -5,7 +5,7 @@
   ******************************************************************************
   * @attention
   *
-  * <h2><center>&copy; COPYRIGHT(c) 2021 STMicroelectronics</center></h2>
+  * <h2><center>&copy; COPYRIGHT(c) 2022 STMicroelectronics</center></h2>
```
Comment (pasitlaothamatas) [inline boards/FSM/Inc/stm32f3xx_hal_conf.h:8]: happy new year

**Comment by pasitlaothamatas** [comment]
notice some warnings here, would you be able to resolve?
image

**Comment by markomarkusi** [comment]
> is there a way to make the jobs appear like this? i think its a little bit neater
> 
> image
> 
> 

Ya prefer this as well if this is possible, otherwise LGTM🐻

File: .github/workflows/githubactions.yml (hunk 3 above)
Comment (akrantz01) [reply .github/workflows/githubactions.yml:208]: It's using the same version that was in the TravisCI scripts. I just copied the surrounding comments as well.

**Comment by akrantz01** [comment]
There isn't much I can do wrt how the checks are displayed since it's a side effect of using the matrix strategy within jobs.

**Comment by pasitlaothamatas** [comment]
any updates on this ? @akrantz01

**Comment by akrantz01** [comment]
There's some weirdness going on with pipenv that I've yet to figure out, hence the recently failing commits

**Comment by akrantz01** [comment]
Rebased to master

**Comment by akrantz01** [comment]
Rebased to master

**Comment by pasitlaothamatas** [comment]
> There isn't much I can do wrt how the checks are displayed since it's a side effect of using the matrix strategy within jobs.

@akrantz01  hmm i see. i do think that going with the matrix is the right approach. Any way to shorten the name of the CubeMX code gen check so we can actually see the board name ?

**Review by pasitlaothamatas — APPROVED** [review]

**Pull request merged**